### PR TITLE
[Runtime] Speed up the long refcounting tests by refcounting in bulk.

### PR DESCRIPTION
**Pull request by mikeash**
The interesting edge cases are always near a power of two. Skip over the uninteresting in-between areas using _n operations, and do single refcounting operations near powers of two to shake out any issues with overflows. Weak ref operations didn't have a _n equivalent, so add that capability to increment/decrementWeak.

### CODE REVIEW

File: unittests/runtime/LongTests/LongRefcounting.cpp [hunk 1]
```diff
@@ -126,24 +129,95 @@ static TestObject *allocTestObject(size_t *addr, size_t value) {
 // Max strong retain count and overflow checking //
 ///////////////////////////////////////////////////
 
+// Perform most of a range with a single _n operation, switching to individual
+// operations within this many steps of a power of two. Inline counts overflow
+// into the side table at a power of two.
```
Comment (Catfish-Man) [inline unittests/runtime/LongTests/LongRefcounting.cpp]: I think the extra check in isOverflowingUnownedRefCount means this isn't strictly true?

File: unittests/runtime/LongTests/LongRefcounting.cpp [hunk 2]
```diff
@@ -341,23 +421,46 @@ TEST(LongRefcountingTest, nonatomic_unowned_retain_overflow_DeathTest) {
...
+
+static void weakReleaseNonAtomic_n(HeapObjectSideTableEntry *side, uint32_t n) {
+  side->decrementWeakNonAtomic(n);
+}
+
 static HeapObjectSideTableEntry *weakRetainALot(TestObject *object, uint64_t count) {
   if (count == 0) return nullptr;
-  
+
   auto side = object->refCounts.formWeakReference();
```
Comment (Catfish-Man) [inline unittests/runtime/LongTests/LongRefcounting.cpp:452]: I assume compensating for stray +1s like this are why there's a +/- 4 window around the power of 2?

**Review by Catfish-Man — APPROVED** [review]
Unnerving to be changing how refcounting works just for the sake of a test, but looks good to me. Left a few minor comments

File: unittests/runtime/LongTests/LongRefcounting.cpp (hunk 1 above)
Comment (mikeash) [reply unittests/runtime/LongTests/LongRefcounting.cpp]: Right, this should say, _near_ a power of two, and that's why there's a window around the powers of two.

**Comment by mikeash** [comment]
@swift-ci please test

File: unittests/runtime/LongTests/LongRefcounting.cpp [hunk 3]
```diff
@@ -341,23 +421,46 @@ TEST(LongRefcountingTest, nonatomic_unowned_retain_overflow_DeathTest) {
...
 static void weakReleaseALot(HeapObjectSideTableEntry *side, uint64_t count) {
-  for (uint64_t i = 0; i < count; i++) {
-    if (atomic) side->decrementWeak();
-    else side->decrementWeakNonAtomic();
-  }
+  if (atomic)
+    runDownward(side, count, weakRelease, weakRelease_n, checkNothing);
+  else
+    runDownward(side, count, weakReleaseNonAtomic, weakReleaseNonAtomic_n,
+                checkNothing);
```
Comment (compnerd) [inline unittests/runtime/LongTests/LongRefcounting.cpp:464]: Why not use ternaries?


```suggestion
  runDownward(side, count,
              atomic ? weakRelease : weakReleaseNonAtomic,
              atomic ? weakRelease_n : weakReleaseNonAtomic_n,
              checkNothing);
```

File: unittests/runtime/LongTests/LongRefcounting.cpp [hunk 4]
```diff
@@ -221,22 +295,28 @@ TEST(LongRefcountingTest, nonatomic_retain_overflow_DeathTest) {
...
-    if (atomic) swift_unownedRelease(object);
-    else swift_nonatomic_unownedRelease(object);
-    EXPECT_ALLOCATED(object);
-  }
+  if (atomic)
+    runDownward(object, count, swift_unownedRelease, swift_unownedRelease_n,
+                checkAllocated);
+  else
+    runDownward(object, count, swift_nonatomic_unownedRelease,
+                swift_nonatomic_unownedRelease_n, checkAllocated);
```
Comment (compnerd) [inline unittests/runtime/LongTests/LongRefcounting.cpp:320]: Similar

File: unittests/runtime/LongTests/LongRefcounting.cpp [hunk 5]
```diff
@@ -221,22 +295,28 @@ TEST(LongRefcountingTest, nonatomic_retain_overflow_DeathTest) {
...
-    if (atomic) swift_unownedRetain(object);
-    else swift_nonatomic_unownedRetain(object);
-    EXPECT_ALLOCATED(object);
-  }
+  if (atomic)
+    runUpward(object, count, swift_unownedRetain, swift_unownedRetain_n,
+              checkAllocated);
+  else
+    runUpward(object, count, swift_nonatomic_unownedRetain,
+              swift_nonatomic_unownedRetain_n, checkAllocated);
```
Comment (compnerd) [inline unittests/runtime/LongTests/LongRefcounting.cpp:310]: Similar

File: unittests/runtime/LongTests/LongRefcounting.cpp [hunk 6]
```diff
@@ -126,24 +129,95 @@ static TestObject *allocTestObject(size_t *addr, size_t value) {
...
-    if (atomic) swift_release(object);
-    else swift_nonatomic_release(object);
-    EXPECT_EQ(0u, deinited);
-  }
+  auto check = [&](TestObject *) { EXPECT_EQ(0u, deinited); };
+  if (atomic)
+    runDownward(object, count, swift_release, swift_release_n, check);
+  else
+    runDownward(object, count, swift_nonatomic_release,
+                swift_nonatomic_release_n, check);
```
Comment (compnerd) [inline unittests/runtime/LongTests/LongRefcounting.cpp:221]: Similar

File: unittests/runtime/LongTests/LongRefcounting.cpp [hunk 7]
```diff
@@ -126,24 +129,95 @@ static TestObject *allocTestObject(size_t *addr, size_t value) {
...
-    if (atomic) swift_retain(object);
-    else swift_nonatomic_retain(object);
-    EXPECT_EQ(0u, deinited);
-  }
+  auto check = [&](TestObject *) { EXPECT_EQ(0u, deinited); };
+  if (atomic)
+    runUpward(object, count, swift_retain, swift_retain_n, check);
+  else
+    runUpward(object, count, swift_nonatomic_retain, swift_nonatomic_retain_n,
+              check);
```
Comment (compnerd) [inline unittests/runtime/LongTests/LongRefcounting.cpp:210]: Similar

File: unittests/runtime/LongTests/LongRefcounting.cpp (hunk 3 above)
Comment (mikeash) [reply unittests/runtime/LongTests/LongRefcounting.cpp:464]: Some of them are preservemost and so aren't the same static type. I'd prefer to have them all be if/else like this rather than having some be ternaries and others not.

**Comment by mikeash** [comment]
@swift-ci please test
